NVDAObjects.IAccessible.sysListView32: ctypes.c_void_p -> ctypes.c_ulong for 32-bit LVItem and LVColumn structures

josephsl opened the pull request:
64-bit transition:

### Link to issue number:
Closes #18706 

### Summary of the issue:
64-bit NVDA cannot obtain column content for 32-bit SysListView32 controls when (re)started while focused on them.

### Description of user facing changes:
64-bit NVDA will announce 32-bit SysListView32 column content when (re)started. A good exmaple is StationPlaylist Studio interface (32-bit app).

### Description of developer facing changes:
In 32-bit LVItem and LVColumn structures, changed "pSZText" field from ctypes.c_void_p to ctypes.c_ulong.

### Description of development approach:
Compared 32-bit and 64-bit LVItem and LVColumn structures to make sure the correct C type was chosen.

### Testing strategy:
Manual: tested 32-bit and 64-bit NVDA wihle using 32-bit apps with SysListView32 controls defined.

### Known issues with pull request:
None

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

josephsl committed:
NVDAObjects.IAccessible.sysListView32: update copyright header. Re #18706

josephsl committed:
SysListView32 structures: ctypes.c_void_p -> ctypes.c_ulong. Re #18706.

64-bit NVDA cannot obtain column data for 32-bit SysListView32 controls when (re)starting NVDA, traced to c_void_p definition in 32-bit structure member definitions. Therefore, edit both the LVITEM and LVCOLUMN structures to use c_ulong (technically an unsigned int) for defining psztext member.

LeonarddeR commented on source/NVDAObjects/IAccessible/sysListView32.py:
```diff
@@ -95,7 +95,7 @@ class LVITEM(Structure):  # noqa: F405
 		("state", c_uint),  # noqa: F405
 		("stateMask", c_uint),  # noqa: F405
-		("pszText", c_void_p),  # noqa: F405
+		("pszText", c_ulong),  # noqa: F405
```
Please add appropriate comments to explain why it is necessary to diverge from `commctrl.h` definitions here
```suggestion
		# A pointer to a buffer containing the text of the item.
		# Note that the pointer size is dictated by the process that hosts the list item, not the process that fetches the list item information.
		("pszText", c_ulong),  # noqa: F405
```

LeonarddeR commented on source/NVDAObjects/IAccessible/sysListView32.py:
```diff
@@ -133,7 +133,7 @@ class LVCOLUMN(Structure):  # noqa: F405
 		("fmt", c_int),  # noqa: F405
 		("cx", c_int),  # noqa: F405
-		("pszText", c_void_p),  # noqa: F405
+		("pszText", c_ulong),  # noqa: F405
```
Same as above

LeonarddeR reviewed (changes requested):
I wonder how C(++) applications are supposed to retrieve this information cross architecture.

josephsl commented:
Hi, this was done based on 64-bit structure definitions. Thanks.

 

From: Leonard de Ruijter ***@***.***> 
Sent: Saturday, August 30, 2025 12:54 AM
To: nvaccess/nvda ***@***.***>
Cc: Joseph Lee ***@***.***>; Author ***@***.***>
Subject: Re: [nvaccess/nvda] NVDAObjects.IAccessible.sysListView32: ctypes.c_void_p -> ctypes.c_ulong for 32-bit LVItem and LVColumn structures (PR #18839)

 

@LeonarddeR requested changes on this pull request.

I wonder how C(++) applications are supposed to retrieve this information cross architecture.

  _____  

In source/NVDAObjects/IAccessible/sysListView32.py <https://github.com/nvaccess/nvda/pull/18839#discussion_r2311820123> :

> @@ -95,7 +95,7 @@ class LVITEM(Structure):  # noqa: F405
               ("iSubItem", c_int),  # noqa: F405
               ("state", c_uint),  # noqa: F405
               ("stateMask", c_uint),  # noqa: F405
-              ("pszText", c_void_p),  # noqa: F405
+              ("pszText", c_ulong),  # noqa: F405

Please add appropriate comments to explain why it is necessary to diverge from commctrl.h definitions here

⬇️ Suggested change 

-              ("pszText", c_ulong),  # noqa: F405
+              # A pointer to a buffer containing the text of the item.
+              # Note that the pointer size is dictated by the process that hosts the list item, not the process that fetches the list item information.
+              ("pszText", c_ulong),  # noqa: F405
  _____  


In source/NVDAObjects/IAccessible/sysListView32.py <https://github.com/nvaccess/nvda/pull/18839#discussion_r2311820293> :

> @@ -133,7 +133,7 @@ class LVCOLUMN(Structure):  # noqa: F405
               ("mask", c_uint),  # noqa: F405
               ("fmt", c_int),  # noqa: F405
               ("cx", c_int),  # noqa: F405
-              ("pszText", c_void_p),  # noqa: F405
+              ("pszText", c_ulong),  # noqa: F405

Same as above

—
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/18839#pullrequestreview-3170915882> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXEE34AKSEVGW5KFNEX33QFDAFAVCNFSM6AAAAACFGAH4KKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCNZQHEYTKOBYGI> .
You are receiving this because you authored the thread.Message ID: ***@***.***>

josephsl committed:
Merge branch 'master' into sysListView32wow64structure

josephsl committed:
SysListView32 structures: add comment on using ulong instead of c_void_p for 32-bit structures. Re #18706.

Review comment from Leonard de Ruijter: add explanatory note on why ctypes unsigned long is used to fetch item text and column text for 32-bit LVItem/LVolumn structures (pointer size is dictated by the app arcihtecture of the hosting process, not NVDA).
